### ISSUE TITLE
Refreshing of session for a community user goes into infinite loop

### ISSUE THREAD

**Issue by ronplanken** [forcedotcom/SalesforceMobileSDK-Android#1437]
1. Version of Mobile SDK Used: 
SalesforceMobileSDK-Android 5.1.0

2. Issue found in Native App or Hybrid App:
The issue was found in a Hybrid-remote app

3. OS Version:
Android 6.0.0 & Android 7.1.1

4. Device:
Android Nexus 5x emulator running Android 6.0.0 and a physical OnePlus 3 running Android 7.1.1

5. Steps to reproduce:
- Create a new hybrid-remote app using the latest forcedroid 
- Open up the newly created app and only update remoteAccessConsumerKey, oauthRedirectURI and a proper startPage in bootconfig.json.
- Start up the app and change the login server to a custom one that goes directly to a community login page
- Login using a user with the Partner Community Login license. (Did not test with different kinds of licenses) 
- App loads the correct startpage and shows it to the user
- Force close the app (using a swipe)
- Wait for the session to timeout (for testing purposes we set the session timeout to 15 minutes)
- Restart the app, enter the pincode and try to show page. 
- The app will see the session has timed out and try to refresh the session token. Trying to do this will force the app into an infinite loop.

6. Actual behavior:
After a session time out with a community user the app will keep on trying to refresh the session token and stall/show a white screen.

7. Expected Behavior:
The expected behavior is for the app to refresh the session token and keep on loading the requested start page.

8. Error Log:
https://gist.github.com/ronplanken/f8e940a1dcf4c3dc215b6a1ddc2ef907 

(Session id's have been scrubbed)

**Comment by trooper2013** [comment]
@ronplanken I've tried these repro steps and but cant replicate the problem.  What type of application is your remote app? Is it a lightning app for instance?

**Comment by ronplanken** [comment]
Our app is a collection of visualforce pages. We are not using lightning components and our orgs are not using the lightning experience yet.

Regular users seem to work fine, did you try to replicate it using community users?

**Comment by bhariharan** [comment]
@ronplanken Could you debug `SalesforceDroidGapActivity` (specifically the refresh code) to see what's happening?

**Comment by ronplanken** [comment]
After trying to debug the Android SDK I have found the following information:

When the app restarts it will try to load a url as follows:

> https ://hr2dtest1.force.com/onzecommunity/secur/frontdoor.jsp?&sid=00D1a000000L2So[SID]&retURL=**https ://hr2dtest1.force.com/onzecommunity/apex/hr2d__EmployeeInteractionCenter?mobile%3D1**&display=touch

If you look at the bold section, it has the proper retUrl for our landing page. Because the session token has been timed out, it will acquire a new token and then use the function getFrontDoorUrl() in SalesforceDroidGapActivity to decide where to go next. This creates the following url:

> http ://hr2dtest1.force.com/onzecommunity/secur/frontdoor.jsp?&sid=00D1a000000L2So![SIDNEW]&**retURL=/onzecommunity/secur/frontdoor.jsp?sid[SIDOLD]%26retURL%3Dhttps%253A%252F%252Fhr2dtest1.force.com%252Fonzecommunity%252Fapex%252Fhr2d__EmployeeInteractionCenter%253Fmobile%253D1**%26display%3Dtouch&display=touch


If you look at the bold section now, it seems the retUrl it contains is the old link trying to login with the old Session id. 

The following line must be the culprit:
`url = (isAbsUrl ? url : client.getClientInfo().resolveUrl(url).toString());`

For testing purposes I hardcode the retUrl to one of my own, and this seems to work every time I ran my tests. This obviously isn't a fix but maybe you guys have a better idea why and how this happens.

**Comment by bhariharan** [comment]
When the session times out, the community is expected to pass along 2 parameters in the redirect URL - `?ec=301` or `?ec=302`, and `startURL`. We use the `ec` parameter to determine that it's a login redirect. You can see this logic in the `isLoginRedirect()` method in `SalesforceWebViewClientHelper.java`. `retUrl` that's used to load the page after refresh is computed from `startURL` that's given to us by the community when the session expiry redirection occurs. Could you verify that your community is passing along the correct `startURL` parameter? Could you also verify that it's sending back `ec=301` or `ec=302` in the parameters when session times out? These are required.

**Comment by ronplanken** [comment]
The url that is returned with the ec=302 value is as followed:

`https://hr2dtest1.force.com/onzecommunity/login?display=touch&ec=302&inst=1a&startURL=%2Fonzecommunity%2Fsecur%2Ffrontdoor.jsp%3Fsid%3D00D1a000000L2So%25[SID]wQJR.Dv.9EqBdb%26retURL%3Dhttps%253A%252F%252Fhr2dtest1.force.com%252Fonzecommunity%252Fapex%252Fhr2d__EmployeeInteractionCenter%253Fmobile%253D1%26display%3Dtouch
`
This part seems to be ok?

**Comment by bhariharan** [comment]
The issue here is the `startURL` returned from your community -> `%2Fonzecommunity%2Fsecur%2Ffrontdoor.jsp%3Fsid%3D00D1a000000L2So%25[SID]wQJR.Dv.9EqBdb%26retURL%3Dhttps%253A%252F%252Fhr2dtest1.force.com%252Fonzecommunity%252Fapex%252Fhr2d__EmployeeInteractionCenter%253Fmobile%253D1%26display%3Dtouch`. Your `startURL` is the `frontdoor` URL, whereas it should be the page you're navigating to. Basically, it should be the same format as the `startURL` you see in your org. You should be able to customize the session timeout redirect URL in your community settings to fix this. There's nothing we can do on the client side for this.

**Comment by ronplanken** [comment]
@bhariharan Thanks for the suggestion. After your latest comment we tried to find out what was going wrong. It seems in the CommunitiesLoginController that is created by Salesforce when you activate communities the startUrl is always directly placed in the new startUrl. This seems to cause a loop in the redirecting after getting a new session token when frontdoor.jsp is used.

We changed our controller to be a bit more intelligent (to make sure a loop won't happen) but this makes me wonder if we really are the only ones that have this issue?
